sys/kern/signal.c

```diff
@@ -201,6 +204,88 @@ int do_sigsuspend(proc_t *p, const sigset_t *mask) {
   return ERESTARTNOHAND;
 }
 
+int do_sigtimedwait(proc_t *p, sigset_t waitset, ksiginfo_t *kinfo,
+                    timespec_t *tsp) {
+  int timevalid = 0, error = 0, timeout = 0;
+  sigset_t saved_mask;
+  signo_t sig;
+  thread_t *td = p->p_thread;
+
+  if (tsp != NULL) {
+    if (tsp->tv_nsec < 0 || tsp->tv_nsec >= 1000000000) {
+      return EINVAL;
+    }
+    /* This is the same behaviour as in OpenBSD. */
+    if (tsp->tv_sec < 0) {
+      return EAGAIN;
+    }
```

Comment:
I think you should only set the `timevalid` variable. Timeout value is chcecked only if you have to wait for signal (according to [FreeBSD manual](https://www.freebsd.org/cgi/man.cgi?query=sigtimedwait&sektion=2&n=1#:~:text=%5BEINVAL%5D%09%09The%20timeout%20argument%20specified%20a%20tv_nsec%20value%20less%0A%09%09%09than%20zero%20or%20greater%20than%20or%20equal%20to%201000%20million.%0A%09%09%09Kernel%20only%20checks%20for%20this%20error%20if%20no%09signal%20is%0A%09%09%09pending%09in%20set%20and%20it%20is%20necessary%20to%20wait.)).

Reply:
I looked both on FreeBSD and OpenBSD implementations and I think I confused them here. However I think that freebsd's manual does not specify in which order it processes when the timeout is invalid:

`If the timespec structure pointed to by timeout is zero-valued and if none
 of the signals specified by set are pending, then sigtimedwait() returns im‐
 mediately with an error.  If timeout is the NULL pointer, the behavior is
 unspecified.` 

In OpenBSD it returns immediately if timeout is invalid or `tsp->tv_sec < 0`: https://mimiker.ii.uni.wroc.pl/source/xref/NetBSD/sys/kern/sys_sig.c?r=53d22107#752

The question now is which approach should we take? Wdyt?

Reply:
In NetBSD manual there is almost identical behavior described as in FreeBSD ([link](https://man.netbsd.org/sigtimedwait.2)). I haven't found this syscall in OpenBSD manual. I'd suggest to stick to the behavior described in manuals. @cahirwpz What do you think?

Reply:
Okay, I think that NetBSD has a little bit better described semantics here (https://man.netbsd.org/sigtimedwait.2). The say that if `timeout == NULL` then it behaves like `sigwaitinfo()` (waits for signal without the timeout). In that case we don't have any problem with order of operations:
1. Check if there are pending signals.
2. Check the timeout value (and return error or wait specified amount of time)

Reply:
Again, I think that the NetBSD manual also do not describe what happens if the timeout value is invalid. They very explicitly say `if timeout is zero [...]` and `if timeout is NULL [...]`.

I'd expect a sentence like this: 'if the timeout value is invalid (i.e. it's false that `0 <= tv_nsec <= 1e9)` or tv_sec is negative, first check current pending signals, if there's no signal that we wait for then return an error.'.

Reply:
As you can see [here](https://mimiker.ii.uni.wroc.pl/source/xref/NetBSD/sys/kern/sys_sig.c?r=53d22107#752), NetBSD returns immediately if timeout is invalid or tv_sec < 0. They do not check the pending signal mask in this case.

Reply:
Okay, It seems that manual may be outdated (I mean the commend about `EINVAL` in [error section](https://man.netbsd.org/sigtimedwait.2#:~:text=This%20error%20is%20only%20checked%20if%20no%20signal%20from%20set%20is%20pending%20and%20it%20would%0A%20%20%20%20%20be%20necessary%20to%20wait.).).

Maybe it is better to filter wrong arguments before anything is checked. It forces user to always pass correct arguments. I think we can leave it like that.